ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -0,0 +1,9955 @@
+|		LOAD_ADDR FCARG2a, filename
+|		.if X64
+|			mov CARG3d, opline->lineno
+|			xor CARG4, CARG4
+|			xor CARG5, CARG5
+|		.else
+|			sub r4, 4
+|			push 0
+|			push 0
+|			push opline->lineno
+|		.endif
+|		EXT_CALL _efree, r0
+|		.if not(X64)
+|			add r4, 4
+|		.endif
+||#else
+|		EXT_CALL _efree_24, r0
+||#endif
+|.endmacro
+
+|.macro EFREE_24, ptr, op_array, opline
+|	mov FCARG1a, ptr
+|	EFREE_REG_24 op_array, opline
+|.endmacro
+
+|.macro EMALLOC, size, op_array, opline
+||#if ZEND_DEBUG
+||		const char *filename = op_array->filename ? op_array->filename->val : NULL;
+|		mov FCARG1a, size
+|		LOAD_ADDR FCARG2a, filename
+|		.if X64
+|			mov CARG3d, opline->lineno
+|			xor CARG4, CARG4
+|			xor CARG5, CARG5
+|		.else
+|			sub r4, 4
+|			push 0
+|			push 0
+|			push opline->lineno
+|		.endif
+|		EXT_CALL _emalloc, r0
+|		.if not(X64)
+|			add r4, 4
+|		.endif
+||#else
+||	if (size == 24) {
+|		EXT_CALL _emalloc_24, r0
+||	} else {
+|		mov FCARG1a, size
+|		EXT_CALL _emalloc, r0
+||	}
+||#endif
+|.endmacro
+
+|.macro OBJ_RELEASE, reg, tmp_reg, exit_label
+|	GC_DELREF reg
+|	jne >1
+|	// zend_objects_store_del(obj);
+|	mov FCARG1a, reg
+|	EXT_CALL zend_objects_store_del, r0
+|	jmp exit_label
+|1:
+|	IF_GC_MAY_NOT_LEAK reg, tmp_reg, >1
+|	// gc_possible_root(obj)
+|	mov FCARG1a, reg
+|	EXT_CALL gc_possible_root, r0
+|1:
+|.endmacro
+
+|.macro UNDEFINED_OFFSET, opline
+||	if (opline == last_valid_opline) {
+|		call ->undefined_offset_ex
+||	} else {
+|		SAVE_VALID_OPLINE, opline
+|		call ->undefined_offset
+||	}
+|.endmacro
+
+|.macro UNDEFINED_INDEX, opline
+||	if (opline == last_valid_opline) {
+|		call ->undefined_index_ex
+||	} else {
+|		SAVE_VALID_OPLINE, opline
+|		call ->undefined_index
+||	}
+|.endmacro
+
+|.macro CANNOT_ADD_ELEMENT, opline
+||	if (opline == last_valid_opline) {
+|		call ->cannot_add_element_ex
+||	} else {
+|		SAVE_VALID_OPLINE, opline
+|		call ->cannot_add_element
+||	}
+|.endmacro
+
+static zend_bool reuse_ip;
+static zend_bool delayed_call_chain;
+static uint32_t  delayed_call_level;
+static const zend_op *last_valid_opline;
+static int jit_return_label;
+
+/* bit helpers */
+
+/* from http://aggregate.org/MAGIC/ */
+static uint32_t ones32(uint32_t x)
+{
+	x -= ((x >> 1) & 0x55555555);
+	x = (((x >> 2) & 0x33333333) + (x & 0x33333333));
+	x = (((x >> 4) + x) & 0x0f0f0f0f);
+	x += (x >> 8);
+	x += (x >> 16);
+	return x & 0x0000003f;
+}
+
+static uint32_t floor_log2(uint32_t x)
+{
+	x |= (x >> 1);
+	x |= (x >> 2);
+	x |= (x >> 4);
+	x |= (x >> 8);
+	x |= (x >> 16);
+	return ones32(x) - 1;
+}
+
+static zend_bool is_power_of_two(uint32_t x)
+{
+	return !(x & (x - 1));
+}
+
+static zend_bool has_concrete_type(uint32_t value_type)
+{
+	if (value_type & MAY_BE_UNDEF) {
+		value_type |= MAY_BE_NULL;
+	}
+	value_type &= MAY_BE_ANY;
+	return is_power_of_two (value_type);
+}
+
+static uint32_t concrete_type(uint32_t value_type)
+{
+	return floor_log2(value_type & MAY_BE_ANY);
```

Comment:
We should probably add some portability wrappers for `__builtin_ctz`. We already use `__builtin_clz`, but not this one.

Reply:
This is portable C code, taken from "Hackers Delight" book.
In any case, I agree to add potable bit scanning wrappers.